### PR TITLE
Add kn ConsoleCLIDownload for disconnected install

### PR DESCRIPTION
**Pull request by navidshaikh**
- Add kn-downloads deployment, service and route using kn-cli-artifacts
 image.
 - Create namespace `serverless-operator` and deploy mentioned resources in it.
 - The deployment runs a container namely; kn-downloads-server. This container
 runs a simple http server using python2, serving the kn cross platform
 artifacts in tar.gz or zip format.
 - Paths for respective platform downloads:
    - /amd64/linux/kn-linux-amd64.tar.gz
    - /amd64/linux/kn-macos-amd64.tar.gz
    - /amd64/linux/kn-windows-amd64.zip
- Removed kn ConsoleCLIDownload manifest, now populates it in the code.

### CODE REVIEW

File: knative-operator/deploy/resources/kn-downloads.yaml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        effect: "NoExecute"
+        tolerationSeconds: 120
+      - key: "node.kubernetes.io/not-ready"
+        operator: "Exists"
+        effect: "NoExecute"
+        tolerationSeconds: 120
+      containers:
+      - name: kn-download-server
+        terminationMessagePolicy: FallbackToLogsOnError
+        image: quay.io/openshift-knative/kn-cli-artifacts:latest
```
Comment (navidshaikh) [inline knative-operator/deploy/resources/kn-downloads.yaml]: using quay based image for testing/review; this will be replaced later

File: knative-operator/deploy/resources/kn-downloads.yaml [hunk 2]
```diff
@@ -0,0 +1,155 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+ name: serverless-operator
```
Comment (navidshaikh) [inline knative-operator/deploy/resources/kn-downloads.yaml]: need suggestion if it should use this namespace

File: knative-operator/deploy/resources/kn-downloads.yaml [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        operator: Exists
+        effect: "NoSchedule"
+      - key: "node.kubernetes.io/unreachable"
+        operator: "Exists"
+        effect: "NoExecute"
+        tolerationSeconds: 120
+      - key: "node.kubernetes.io/not-ready"
+        operator: "Exists"
+        effect: "NoExecute"
+        tolerationSeconds: 120
```
Comment (markusthoemmes) [inline knative-operator/deploy/resources/kn-downloads.yaml]: All the tolerations are redundant I think. We shouldn't specify them.

File: knative-operator/deploy/resources/kn-downloads.yaml [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+      component: kn-downloads
+  template:
+    metadata:
+      name: kn-downloads
+      labels:
+        app: serverless-operator
+        component: kn-downloads
+    spec:
+      nodeSelector:
+        kubernetes.io/os: "linux"
```
Comment (markusthoemmes) [inline knative-operator/deploy/resources/kn-downloads.yaml]: Why is this needed?

File: knative-operator/deploy/resources/kn-downloads.yaml [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        operator: "Exists"
+        effect: "NoExecute"
+        tolerationSeconds: 120
+      - key: "node.kubernetes.io/not-ready"
+        operator: "Exists"
+        effect: "NoExecute"
+        tolerationSeconds: 120
+      containers:
+      - name: kn-download-server
+        terminationMessagePolicy: FallbackToLogsOnError
```
Comment (markusthoemmes) [inline knative-operator/deploy/resources/kn-downloads.yaml]: This is the default value, let's not specify it.

File: knative-operator/deploy/resources/kn-downloads.yaml (hunk 2 above)
Comment (markusthoemmes) [reply knative-operator/deploy/resources/kn-downloads.yaml]: I'm wondering if we can make this part of the CSV instead 🤔 . @jcrossley3 wdyt?

File: knative-operator/deploy/resources/kn-downloads.yaml [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            path: /
+            port: 8080
+            scheme: HTTP
+        command: ["/bin/sh"]
+        resources:
+          requests:
+            cpu: 10m
+            memory: 50Mi
+        args:
+        - '-c'
```
Comment (markusthoemmes) [inline knative-operator/deploy/resources/kn-downloads.yaml]: Wouldn't this be easier with an nginx image or something like that?

Comment (navidshaikh) [reply knative-operator/deploy/resources/kn-downloads.yaml]: Yes, however I re-used most of these specs from `oc` `ConsoleCLIDownload which is being shipped/tested. We can refine this in next iterations once we've this functioning end to end.

File: knative-operator/deploy/resources/kn-downloads.yaml (hunk 5 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/kn-downloads.yaml]: please correct me if I misunderstood, but I see default value for this field as `File`
https://v1-16.docs.kubernetes.io/docs/tasks/debug-application-cluster/determine-reason-pod-failure/

File: knative-operator/deploy/resources/kn-downloads.yaml (hunk 4 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/kn-downloads.yaml]: again, coming as is from oc spec, will remove it.

File: knative-operator/deploy/resources/kn-downloads.yaml (hunk 3 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/kn-downloads.yaml]: Removed.

File: knative-operator/deploy/resources/kn-downloads.yaml (hunk 2 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/kn-downloads.yaml]: @jcrossley3 ping

Comment (jcrossley3) [reply knative-operator/deploy/resources/kn-downloads.yaml]: I don't understand the question -- you want to make a namespace part of the CSV?

Comment (navidshaikh) [reply knative-operator/deploy/resources/kn-downloads.yaml]: There are two questions here:

1 - Which namespace to use to create this deployment, route, service into?
2 - Where to put all these resources YAMLs?

File: hack/install.sh [hunk 7]
```diff
@@ -13,7 +13,7 @@ source "$(dirname "${BASH_SOURCE[0]}")/lib/__sources__.bash"
 
 set -Eeuo pipefail
 
-scale_up_workers || exit $?
+#scale_up_workers || exit $?
```
Comment (navidshaikh) [inline hack/install.sh]: TODO: remove

**Comment by bbrowning** [comment]
Why are we creating a new namespace just for the CLI download pod? Can we not put it one of our existing namespaces - either the one the Serverless Operator is installed in, or the knative-serving namespace?

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            scheme: HTTP
+        command: ["/bin/sh"]
+        resources:
+          requests:
+            cpu: 10m
+            memory: 50Mi
+        args:
+        - '-c'
+        - |
+          cat <<EOF >>/tmp/serve.py
```
Comment (bbrowning) [inline knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml]: This feels like a separate file we should version control instead of an inline python script here.

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,141 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+ name: serverless-operator
+---
+apiVersion: apps/v1
+kind: Deployment
```
Comment (bbrowning) [inline knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml]: Should this be a Knative Service instead of a Deployment? So that it scales to 0 when not in use?

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    metadata:
+      name: kn-cli-downloads
+      labels:
+        app: serverless-operator
+        component: kn-cli-downloads
+    spec:
+      containers:
+      - name: kn-download-server
+        terminationMessagePolicy: FallbackToLogsOnError
+        image: quay.io/openshift-knative/kn-cli-artifacts:latest
```
Comment (bbrowning) [inline knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml]: Should this ever pull latest? I'd expect it to be pinned to the specific `kn` version that directly matches this release of the operator. And we'll have to ensure automation replaces this with the corresponding actual product build from registry.redhat.io instead of quay.io.

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        imagePullPolicy: Always
```
Comment (bbrowning) [inline knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml]: `imagePullPolicy: Always` can go away once we stop using the latest image tag.

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml (hunk 10 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml:27]: Nope, not latest. It was added for testing (even before we cut any tagged release of this image).
https://github.com/openshift-knative/serverless-operator/pull/178#discussion_r392846395

**Comment by navidshaikh** [comment]
@bbrowning  
> Why are we creating a new namespace just for the CLI download pod? Can we not put it one of our existing namespaces - either the one the Serverless Operator is installed in, or the knative-serving namespace?

I am looking for suggestion on that https://github.com/openshift-knative/serverless-operator/pull/178#discussion_r392847071

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml (hunk 11 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml]: Removed.

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml (hunk 10 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml]: Added tagged version for now from openshift registry.

**Comment by navidshaikh** [comment]
/test 4.4-e2e-aws-ocp-44

        kubelogs.go:157: D 11:34:37.172 [scaling/autoscaler.go:211] [serving-tests/activator-overload-zvpuzejk-2ws5v] Operating in panic mode.

**Comment by navidshaikh** [comment]
/test 4.3-e2e-aws-ocp-43

    Operating in panic mode.

**Comment by navidshaikh** [comment]
New changes to PR:
 - Add kn-cli-artifacts co-ordinates in CSV 
 - Add kn download resources manifest path in CSV
 - Use `kn-cli-artifacts` namespace to deploy resources (earlier namespace created/used was `serverless-operator`)

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 12]
```diff
@@ -1,51 +1,187 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	servingv1alpha1 "knative.dev/serving-operator/pkg/apis/serving/v1alpha1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-const defaultKnConsoleCLIDownload = "deploy/resources/console_cli_download_kn.yaml"
+const (
+	knDownloadServer                    = "kn-download-server"
+	knConsoleCLIDownloadDeployRoute     = "kn-cli-downloads"
+	knConsoleCLIDownloadDeployNamespace = "kn-cli-downloads"
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:28]: As discussed in Slack, I think we agreed on leaving this in knative-serving for now.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 13]
```diff
@@ -1,51 +1,187 @@
...
 
 	"github.com/openshift-knative/serverless-operator/knative-operator/pkg/common"
-	servingv1alpha1 "knative.dev/serving-operator/pkg/apis/serving/v1alpha1"
 
 	mfc "github.com/manifestival/controller-runtime-client"
 	mf "github.com/manifestival/manifestival"
+	consolev1 "github.com/openshift/api/console/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	consoleroute "github.com/openshift/console-operator/pkg/console/subresource/route"
+	consoleutil "github.com/openshift/console-operator/pkg/console/subresource/util"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:16]: I'd thought it while adding import, wasn't sure though, copied over now. Indeed it removed a lot of dep files.

File: knative-operator/deploy/resources/console_cli_download_kn_resources.yaml [hunk 14]
```diff
@@ -0,0 +1,136 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  namespace: placeholder
```
Comment (navidshaikh) [inline knative-operator/deploy/resources/console_cli_download_kn_resources.yaml:4]: using `placeholder` to be replaced by instance's namespace

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 15]
```diff
@@ -1,51 +1,240 @@
...
+	manifest, err := rawManifest(apiclient)
+	if err != nil {
+		return mf.Manifest{}, fmt.Errorf("failed to read kn ConsoleCLIDownload deployment manifest: %w", err)
+	}
+
+	// 1. Use instance's namespace to deploy download resources into
+	// 2. Set proper kn-cli-artifacts image
+	transforms := []mf.Transformer{mf.InjectNamespace(instance.GetNamespace()),
+		replaceKnCLIArtifactsImage(os.Getenv("IMAGE_KN_CLI_ARTIFACTS"), scheme),
+	}
```
Comment (navidshaikh) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:130]: should we add here another transformation which adds annotations for its owner OR ownerReferences ?

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 16]
```diff
@@ -1,51 +1,187 @@
...
+
+	return nil
+}
+
+// createCR creates kn ConsoleCLIDownload CR. It finds the route serving kn binaries and
+// populates ConsoleCLIDownload object accordingly and creates it.
+func createCR(apiclient client.Client) error {
+	route := &routev1.Route{}
+	// Waiting for the deployment/route to appear. TODO: ideally this should be a watch
+	time.Sleep(time.Second * 10)
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go]: Added using `k8s.io/apimachinery/pkg/util/wait.Pollimmediate` with interval of 2 secs and timeout of 20 secs.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 17]
```diff
@@ -1,51 +1,187 @@
...
+func addToScheme(scheme *runtime.Scheme) {
+	scheme.AddKnownTypes(consolev1.GroupVersion, &consolev1.ConsoleCLIDownload{})
+	scheme.AddKnownTypes(k8sv1.SchemeGroupVersion, &k8sv1.Deployment{})
+}
+
+// populateKnConsoleCLIDownload populates kn ConsoleCLIDownload object and its SPEC
+// using route's baseURL
+func populateKnConsoleCLIDownload(baseURL string) *consolev1.ConsoleCLIDownload {
+	return &consolev1.ConsoleCLIDownload{
+		metav1.TypeMeta{
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go]: updated with struct fields name

**Comment by navidshaikh** [comment]
/retest

setup issues

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (markusthoemmes) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go]: That's not a watch though. We cannot block the reconcile loop for longer than maybe a second or so. This will block the reconcile loop for 20seconds worst case.

[the remark this answers is not in the document]
Comment (markusthoemmes) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go]: A watch would need to be setup via setting up a literal `Watch`, listening to the events of the resources and triggering a reconcile once something changes.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go (hunk 15 above)
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go]: Added annotations for owner as done for kourier.

**Comment by navidshaikh** [comment]
/retest

 setup issues

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml (hunk 8 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml]: Added https://github.com/openshift/knative-client/pull/328/ to version control serve.py, and configured it to run as ENTRYPOINT for kn-cli-artifacts image.

File: knative-operator/pkg/controller/knativeserving/knativeserving_controller_test.go [hunk 18]
```diff
@@ -45,12 +48,36 @@ var (
 	defaultRequest = reconcile.Request{
 		NamespacedName: types.NamespacedName{Namespace: "knative-serving", Name: "knative-serving"},
 	}
+
+	defaultKnRoute = routev1.Route{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "ConsoleCLIDownload",
+		},
```
Comment (navidshaikh) [inline knative-operator/pkg/controller/knativeserving/knativeserving_controller_test.go:55]: bad copy paste, will remove.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 19]
```diff
@@ -1,51 +1,310 @@
...
 	return nil
 }
 
-// Delete deletes ConsoleCLIDownload for kn CLI download links
-func Delete(instance *servingv1alpha1.KnativeServing, apiclient client.Client) error {
-	log.Info("Deleting ConsoleCLIDownload CR for kn")
-	manifest, err := mfc.NewManifest(manifestPath(), apiclient, mf.UseLogger(log.WithName("mf")))
+// Check for deployments of kn ConsoleCLIDownload
+// This function is copied from knativeserving_controller.go in serving-operator
+func checkDeployments(manifest *mf.Manifest, instance *servingv1alpha1.KnativeServing, api client.Client) error {
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:71]: nit: `checkResources` maybe? It doesn't only check deployments.

File: knative-operator/pkg/controller/knativeserving/knativeserving_controller.go [hunk 20]
```diff
@@ -95,6 +96,45 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 			return err
 		}
 	}
+
+	// Watch for kn ConsoleCLIDownload resources
+	knToRequestsFunc := handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+		annotations := obj.Meta.GetAnnotations()
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/knativeserving_controller.go:102]: This seems to be a copy of the function used above in kourier. Should we unify the used constants and define the function ony once?

File: knative-operator/pkg/controller/knativeserving/knativeserving_controller.go [hunk 21]
```diff
@@ -95,6 +96,45 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+	})
+
+	knManifest, err := consoleclidownload.RawManifest(mgr.GetClient())
+	if err != nil {
+		return err
+	}
+	knResources := knManifest.Resources()
+
+	gvkToCCD := make(map[schema.GroupVersionKind]runtime.Object)
+	for i := range knResources {
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/knativeserving_controller.go:120]: Let's add the CCD itself to this list too. Changes to that should be reconciled by us as well, right?

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 22]
```diff
@@ -1,51 +1,310 @@
...
+func https(host string) string {
+	protocol := "https://"
+	if host == "" {
+		return ""
+	}
+	if strings.HasPrefix(host, protocol) {
+		return host
 	}
-	return knConsoleCLIDownload
+	secured := fmt.Sprintf("%s%s", protocol, host)
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:308]:
```suggestion
	return protocol + host
```

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 23]
```diff
@@ -1,51 +1,310 @@
...
+		if condition.Type == routev1.RouteAdmitted && condition.Status == corev1.ConditionTrue {
+			admitted = true
+		}
+	}
+	return admitted
+}
+
+// copied from github.com/openshift/console-operator/pkg/console/subresource/util/util.go and modified
+func https(host string) string {
+	protocol := "https://"
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:301]: nit: define this after the empty check.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 24]
```diff
@@ -1,51 +1,310 @@
...
+// Function copied from github.com/openshift/console-operator/pkg/console/subresource/route/route.go
+func isIngressAdmitted(ingress routev1.RouteIngress) bool {
+	admitted := false
+	for _, condition := range ingress.Conditions {
+		if condition.Type == routev1.RouteAdmitted && condition.Status == corev1.ConditionTrue {
+			admitted = true
+		}
+	}
+	return admitted
+}
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:297]:
```suggestion
func isIngressAdmitted(ingress routev1.RouteIngress) bool {
	for _, condition := range ingress.Conditions {
		if condition.Type == routev1.RouteAdmitted && condition.Status == corev1.ConditionTrue {
			return true
		}
	}
	return false
}
```

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 25]
```diff
@@ -1,51 +1,310 @@
...
+				},
+			},
+		},
+	}
+}
+
+// Function copied from github.com/openshift/console-operator/pkg/console/subresource/route/route.go and modified
+func getCanonicalHost(route *routev1.Route) string {
+	for _, ingress := range route.Status.Ingress {
+		if ingress.RouterName != defaultIngressController {
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:276]: Why do we care about this check? Does this mean that this won't work with non-default ingress controllers?

I feel like we should just return the first non-empty `ingress.Host` 🤷‍♂️

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 26]
```diff
@@ -1,51 +1,310 @@
...
+
+	// Check if kn ConsoleCLIDownload exists
+	err := apiclient.Get(context.TODO(), client.ObjectKey{Namespace: "", Name: knCLIDownload}, knConsole)
+	switch {
+	case apierrors.IsNotFound(err):
+		log.Info("Creating kn ConsoleCLIDownload")
+		if err := apiclient.Create(context.TODO(), knConsoleObj); err != nil {
+			return err
+		} else {
+			// handle upgrade scenario: In 1.6 we'd created kn ConsoleCLIDownload CO with
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:124]: Why are we changing the name? We should be able to just reconcile in the needed changes I think?

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 27]
```diff
@@ -1,51 +1,310 @@
...
+			if err := apiclient.Get(context.TODO(), client.ObjectKey{Namespace: "", Name: "kn"}, knConsoleOld); err == nil {
+				log.Info("Found old kn ConsoleCLIDownload CO 'kn', deleting it..")
+				if err := apiclient.Delete(context.TODO(), knConsoleOld); err != nil {
+					log.Info("failed to delete earlier kn ConsoleCLIDownload CO 'kn' %w", err)
+				}
+				log.Info("Deleted old kn ConsoleCLIDownload CO 'kn'")
+			}
+
+		}
+	case apierrors.IsAlreadyExists(err):
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:138]: This will never happen. A `Get` never returns a `IsAlreadyExists` error.

Instead, this function should be called `reconcileKnConsoleCLIDownload` and follow the layout of other reconcile functions. `reconcileConfigMap` for example, though this can be much less complex.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go [hunk 28]
```diff
@@ -1,51 +1,310 @@
...
+	log.Info("Installing kn ConsoleCLIDownload")
+	route := &routev1.Route{}
+	knConsole := &consolev1.ConsoleCLIDownload{}
+
+	// find the route first
+	if err := apiclient.Get(context.TODO(), client.ObjectKey{Namespace: instance.GetNamespace(), Name: knConsoleCLIDownloadDeployRoute}, route); err != nil {
+		return fmt.Errorf("failed to find kn ConsoleCLIDownload deployment route")
+	}
+
+	knRoute := getCanonicalHost(route)
```
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:113]: Don't we need to handle an empty route here?

File: knative-operator/pkg/controller/knativeserving/knativeserving_controller.go (hunk 20 above)
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/knativeserving_controller.go:102]: I was inclined to do this but waited for feedback before changing kourier bits.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go (hunk 28 above)
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:113]: The empty route is handled in `checkDeployments`, which ensures routes gets a value before this function `applyKnConsoleCLIDownload` is invoked.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go (hunk 26 above)
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:124]: Changed its name from `kn` to `kn-cli-downloads` (following how its done for `oc`).
However, I think there is repetition, no need to explicitly name `cli-downloads`  to ConsoleCLIDownloads CO. A specific name should suffice. I will revert name change and respective code to handle it.

File: knative-operator/pkg/controller/knativeserving/knativeserving_controller.go (hunk 20 above)
Comment (markusthoemmes) [reply knative-operator/pkg/controller/knativeserving/knativeserving_controller.go:102]: We can also do this in a followup.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go (hunk 25 above)
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:276]: I am not sure about this, I copied it over to avoid vendoring. I will remove it for now and lets see how it goes ;-)

File: knative-operator/deploy/resources/kn-downloads.yaml (hunk 2 above)
Comment (bbrowning) [reply knative-operator/deploy/resources/kn-downloads.yaml]: We should definitely not create a new namespace. In the absence of any better ideas, it seems reasonable to me to just put this Deployment into the CSV and then OLM will automatically install it into whichever namespace the operator is installed into.

Comment (navidshaikh) [reply knative-operator/deploy/resources/kn-downloads.yaml]: The latest on the PR applies Deployment, Route and Service in KnativeServing instance's namespace.

File: knative-operator/pkg/controller/knativeserving/knativeserving_controller.go (hunk 21 above)
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/knativeserving_controller.go:134]: Yup, I'd added a separate watch for CCD, appended CCD in this list now.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go (hunk 27 above)
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go]: Right.
Thanks for the ref, updated following `reconcileConfigMap`.

File: knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go (hunk 28 above)
Comment (navidshaikh) [reply knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go:109]: Re-verifying the route URL value now before populating CCD.

**Comment by navidshaikh** [comment]
TODO: Add tests.

File: test/service.go [hunk 29]
```diff
@@ -108,6 +108,35 @@ func CheckDeploymentScale(ctx *Context, ns, name string, scale int) error {
 	return nil
 }
 
+func RouteHost(ctx *Context, ns, name string) (string, error) {
```
Comment (markusthoemmes) [inline test/service.go:111]:
```suggestion
func CheckRouteIsReady(ctx *Context, ns, name string) (string, error) {
```

File: test/e2e/knative_serving_test.go [hunk 30]
```diff
@@ -73,6 +73,25 @@ func TestKnativeServing(t *testing.T) {
...
+		// Check the status of Deployment for kn ConsoleCLIDownload
+		if err := test.CheckDeploymentScale(caCtx, knativeServing, "kn-cli-downloads", "1"); err != nil {
+			t.Fatalf("failed to verify kn ConcoleCLIDownload deployment: %v", err)
+		}
+		// Verify that Route for kn ConsoleCLIDownload is ready and has a host
+		host, err := test.RouteHost(caCtx, knativeServing, "kn-cli-downloads")
+		if err != nil {
+			t.Fatalf("failed to verify kn ConcoleCLIDownload route: %v", err)
+		}
+		if host == "" {
```
Comment (markusthoemmes) [inline test/e2e/knative_serving_test.go:86]: This can never happen, I believe?

File: test/service.go [hunk 31]
```diff
@@ -108,6 +108,35 @@ func CheckDeploymentScale(ctx *Context, ns, name string, scale int) error {
...
+		return "", err
+	}
+	ready, _ := RouteHasHost(r, nil)
+	if !ready {
+		return "", fmt.Errorf("route %s/%s is not ready yet")
+	}
+	return r.Status.Ingress[0].Host, nil
+}
+
+func VerifyConcoleCLIDownload(ctx *Context, name, host string) error {
```
Comment (markusthoemmes) [inline test/service.go:123]: Why not inline this check? I don't see us reusing it.

File: test/e2e/knative_serving_test.go [hunk 32]
```diff
@@ -73,6 +73,25 @@ func TestKnativeServing(t *testing.T) {
...
+			t.Fatalf("kn ConsoleCLIDownload resource route is empty")
+		}
+		// Verify kn ConsoleCLIDownload CO and if download links are cluster local
+		if err := test.VerifyConcoleCLIDownload(caCtx, "kn", host); err != nil {
+			t.Fatalf("failed to verify kn ConsoleCLIDownload: %v", err)
+		}
```
Comment (markusthoemmes) [inline test/e2e/knative_serving_test.go:92]: As a last test, would it be possible to:
1. Call the route and check the HTML contains the links the CCD contains?
2. Try the links themselves and see if they'd result in a download that has more than... dunno, 1MB? How big is kn? 😂 . HEAD requests might work for that.

I'd be fine adding these tests in subsequent PRs.

File: test/e2e/knative_serving_test.go (hunk 30 above)
Comment (navidshaikh) [reply test/e2e/knative_serving_test.go:86]: Yup, if Route is ready its Ingress got to have a `.Host`.

File: test/e2e/knative_serving_test.go (hunk 32 above)
Comment (navidshaikh) [reply test/e2e/knative_serving_test.go:144]: Added checks to verify length of the links, if it starts with the route URL and downloads each to verify > 10MB.

File: test/service.go (hunk 31 above)
Comment (navidshaikh) [reply test/service.go]: inlined.

**Comment by navidshaikh** [comment]
@markusthoemmes : thanks for your patience, ready for another review. :-) 

I'd like squash the commits if its ready to go.

File: knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml (hunk 8 above)
Comment (navidshaikh) [reply knative-operator/deploy/resources/console_cli_download_kn_deployment.yaml]: Removed the cmd/args from deployment manifest, now the file is versioned controlled in openshift/knative-client repo and runs part of image startup.

File: olm-catalog/serverless-operator/1.7.0/serverless-operator.v1.7.0.clusterserviceversion.yaml [hunk 33]
```diff
@@ -376,6 +378,8 @@ spec:
                       value: registry.svc.ci.openshift.org/openshift/knative-v0.13.0:knative-eventing-filter
                     - name: IMAGE_DISPATCHER_IMAGE
                       value: registry.svc.ci.openshift.org/openshift/knative-v0.13.0:knative-eventing-channel-dispatcher
+                    - name: IMAGE_KN_CLI_ARTIFACTS
+                      value: registry.svc.ci.openshift.org/openshift/knative-v0.13.2:kn-cli-artifacts
```
Comment (navidshaikh) [inline olm-catalog/serverless-operator/1.7.0/serverless-operator.v1.7.0.clusterserviceversion.yaml:382]: @warrenvw : please note the co-ordinates we'd need to replace with productized kn-cli-artifacts image.

File: olm-catalog/serverless-operator/1.7.0/serverless-operator.v1.7.0.clusterserviceversion.yaml [hunk 34]
```diff
@@ -531,5 +535,7 @@ spec:
       value: registry.svc.ci.openshift.org/openshift/knative-v0.13.0:knative-eventing-filter
     - name: IMAGE_DISPATCHER_IMAGE
       value: registry.svc.ci.openshift.org/openshift/knative-v0.13.0:knative-eventing-channel-dispatcher
+    - name: IMAGE_KN_CLI_ARTIFACTS
+      value: registry.svc.ci.openshift.org/openshift/knative-v0.13.2:kn-cli-artifacts
```
Comment (navidshaikh) [inline olm-catalog/serverless-operator/1.7.0/serverless-operator.v1.7.0.clusterserviceversion.yaml:539]: @warrenvw : please note the co-ordinates we'd need to replace with productized kn-cli-artifacts image.

File: test/e2e/knative_serving_test.go (hunk 32 above)
Comment (navidshaikh) [reply test/e2e/knative_serving_test.go:117]: updated with to check content-length with HEAD requests, thanks for the suggestion and explaining me :+1:

File: test/e2e/knative_serving_test.go [hunk 35]
```diff
@@ -75,6 +76,45 @@ func TestKnativeServing(t *testing.T) {
...
+		// Verify if individual link starts with correct route
+		protocol := "https://"
+		if !strings.HasPrefix(host, protocol) {
+			host = protocol + host
+		}
+		for _, link := range ccd.Spec.Links {
+			if !strings.HasPrefix(link.Href, host) {
+				t.Fatalf("incorrect href found for kn CCD, expecting prefix %s, found link %s", host, link.Href)
+				h, err := http.Head(link.Href)
```
Comment (markusthoemmes) [inline test/e2e/knative_serving_test.go:109]: This is in the wrong spot I believe? It needs to move out of the if block.

Comment (navidshaikh) [reply test/e2e/knative_serving_test.go:109]: doh :-<

**Comment by markusthoemmes** [comment]
Note: 4.5 is broken on master, likely due to external changes
